### ISSUE TITLE
cmp_owned false positives - proc_macro Ident

### ISSUE THREAD

**Issue by elichai** [rust-lang/rust-clippy#4874]
Hi,
I noticed a false positive when comparing a string with `ident.to_string()`
clippy suggests you don't need to add `to_string()` and you can instead dereference the ident.
1. You can't dereference it because there's no reference involved.
2. there's no `PartialEq` implementation between `&str`/`String` and `Ident`.

Example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b6ae15acb40aee8fe582f1620ac569d7

**Comment by mikerite** [comment]
Your example is:
```
extern crate proc_macro2;
use proc_macro2::*;

fn main() {
    let a1 = Ident::new("Hi", Span::call_site());
    if "Hi" == a1.to_string() {
        println!("yay");
    } else {
        println!(":(");
    }
}
```

proc-macro2 has `impl<T: ?Sized> PartialEq<T> for Ident where T: AsRef<str>` so you can write 
`a1 == "Hi"` to avoid creating the string. 

The reason for the bug is that the lint assumes that `PartialEq` is implemented symmetrically i.e. `T: PartialEq<U>` implies `U: PartialEq<T>`. The [PartialEq](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html) docs actually require this so maybe proc-macro2 is missing an impl.

**Comment by elichai** [comment]
Hi, Nice catch :)
bad I'm not sure the solution you're proposing is even possible: https://play.rust-lang.org/?gist=75e6e15847489b8c6ee7c8af6507dd66
